rbd.c

```diff
@@ -612,7 +616,177 @@ static int tcmu_rbd_lock_break(struct tcmu_device *dev, char **orig_owner)
+		ret = -ENOENT;
+		goto done;
+	}
+
+	metadata_owner = strstr(buf, "rbd_client=");
+	if (!metadata_owner) {
+		tcmu_dev_err(dev, "Invalid lock tag %s.\n", buf);
+		/* Force initaitor to retry STPG */
+		ret = -ENOENT;
+		goto free_owners;
+	}
+
+	metadata_owner += 11;
+	if (strcmp(metadata_owner, owners[0])) {
+		tcmu_dev_dbg(dev, "owner mismatch %s %s\n", metadata_owner,
+			     owners[0]);
+		/*
+		 * A node could be in the middle of grabbing the lock or it
+		 * failed midway. Force tcmu to report all standby so the
+		 * initiator retries the STPG for the failure case.
+		 */
+		ret = -ENOENT;
+		goto free_owners;
+	}
+
+	ret = sscanf(buf, "tcmu_tag=%hu,%*s", tag);
+	if (ret != 1) {
+		tcmu_dev_err(dev, "Invalid lock tag %s.\n", buf);
+		/* Force initaitor to retry STPG */
+		ret = -ENOENT;
+		goto free_owners;
+	}
+	ret = 0;
+
+free_owners:
+	if (num_owners)
+		rbd_lock_get_owners_cleanup(owners, num_owners);
+
+done:
+	switch (ret) {
+	case 0:
+	case -ENOENT:
+	case -ESHUTDOWN:
+	case -ETIMEDOUT:
+		break;
+	default:
+		ret = -EIO;
+		break;
+	}
+
+	return ret;
+}
+
+static int tcmu_rbd_set_lock_tag(struct tcmu_device *dev, uint16_t tcmu_tag)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	rbd_lock_mode_t lock_mode;
+	char *owners[1];
+	size_t num_owners = 1;
+	char *tcmu_rbd_tag;
+	int ret;
+
+	/*
+	 * We cannot take the lock and set the tag atomically. In case
+	 * we fail here and are in an inconsistent state, we attach the rbd
+	 * client lock info along with the tcmu locker tag so remote nodes
+	 * can check the rbd info against rbd_lock_get_owners to determine if
+	 * the tcmu locker tag is current/valid.
+	 */
+	ret = rbd_lock_get_owners(state->image, &lock_mode, owners,
+				  &num_owners);
+	tcmu_dev_dbg(dev, "set tag get lockowner got %d %d\n", ret, num_owners);
+	if (ret)
+		goto done;
+
+	if (!num_owners) {
+		ret = -ENOENT;
+		goto done;
+	}
+
+	ret = asprintf(&tcmu_rbd_tag, TCMU_RBD_LOCKER_TAG_FMT, tcmu_tag,
+		       owners[0]);
+	if (ret < 0) {
+		ret = -EIO;
+		goto free_owners;
+	}
+
+	ret = rbd_metadata_set(state->image, TCMU_RBD_LOCKER_TAG_KEY,
+			       tcmu_rbd_tag);
+	free(tcmu_rbd_tag);
+	if (ret)
+		tcmu_dev_err(dev, "Could not store lock tag. Err %d.\n", ret);
+
+free_owners:
+	if (num_owners)
+		rbd_lock_get_owners_cleanup(owners, num_owners);
+
+done:
+	switch (ret) {
+	case 0:
+	case -ENOENT:
+	case -ESHUTDOWN:
+	case -ETIMEDOUT:
+		break;
+	default:
+		ret = -EIO;
+		break;
+	}
+
+	return ret;
+}
+
+static int tcmu_rbd_rc_to_lock_state(int rc)
+{
+	switch (rc) {
+	case -ESHUTDOWN:
+		return TCMUR_DEV_LOCK_FENCED;
+	case -ETIMEDOUT:
+	default:
+		return TCMUR_DEV_LOCK_UNKNOWN;
+	}
+}
+
+static int tcmu_rbd_unlock(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	int ret;
+
+	if (tcmu_rbd_has_lock(dev) != 1)
+		return TCMUR_DEV_LOCK_UNLOCKED;
+
+	ret = rbd_metadata_remove(state->image, TCMU_RBD_LOCKER_TAG_KEY);
```

Comment:
Is it possible this could race w/ another client stealing the lock? Perhaps just leave it set to the old value and let the mismatch code above detect that it's out-of-date?

Reply:
Ok sounds good. I was thinking if another client took the lock we would be blacklisted by the time we got here. But yeah your suggestion is nicer because we have that other mismatch code and then it saves the worry here. Will drop.